src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -863,13 +870,23 @@ public static object DeepCopyInner(object original)
             {
                 copy = copier(original);
                 SerializationContext.Current.RecordObject(original, copy);
+                return copy;
             }
-            else
+
+            var externalSerializer = externalSerializers.FirstOrDefault(serializer => serializer.IsSupportedType(t));
```

Comment:
@jason-bragg sorry, maybe I misunderstood the proposal... If the idea is to have a ordered list of serializers and the first one able to serialize the object (how will they "know" if they can or can not do it?) will catch it, then I agree with you that it is the recipe for chaos.

I thinking on have all the serializers registered as possible ones regardless of its capabilities or order. Then the same way ou select which StorageProvider you will use for a specific grain, you would select which serializer you want to use for that grain/type.

If we do this per type, lets say, using an attribute, some config file, some register fluent API, DI, whatever the registrar mechanism, you will be able to leverage that serializer for storage providers (when is time to call Grain.WriteStateAsync()/ReadStateAsync()) and when you need to serialize a type to pass over the message to another silo (in case that grain reference is outside Silo boundaries).

I just want to make a single semantics for register and use serializers so I'll make sure whenever my type X is serialized, it will got serialized by the correct registered serializer that I previously choose, instead of have the surprise of my grains data no longer being encrypted 'cause the first serializer on the list does not do it and now can serialize my object as your stated example.